protocol/p2p/src/core/hub.rs

```diff
@@ -85,6 +85,28 @@ impl Hub {
         }
     }
 
+    /// Selects a random subset of peers, trying to select at least half for outbound when possible
+    fn select_some_peers(&self, num_peers: usize) -> Vec<Arc<Router>> {
+        let (outbound_peers, inbound_peers): (Vec<_>, Vec<_>) =
+            self.peers.read().values().cloned().partition(|peer| peer.is_outbound());
+
+        let mut required_outbound_count = (num_peers + 1) / 2;
+
+        // If won't be enough inbound peers to meet the num_peers after we've selected only half for outbound,
+        // try to require more outbound peers for the difference
+        if inbound_peers.len() + required_outbound_count < num_peers {
+            required_outbound_count = num_peers - inbound_peers.len();
+        }
+
+        let thread_rng = &mut rand::thread_rng();
+        // Randomly select about half from outbound
+        let mut peers = outbound_peers.choose_multiple(thread_rng, required_outbound_count).cloned().collect::<Vec<_>>();
```

Comment:
Hmm, it shouldn't panic. If you choose more than there is in the vec, you'd get all the elements (in a random order).

See: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=58d9cdc3cc34466b1dbc19883f82579b

Also, in the code for `choose_multiple`, it will choose the min of the iterable length or the amount passed:

Code: https://rust-random.github.io/rand/src/rand/seq/mod.rs.html#560

```
    #[cfg(feature = "alloc")]
    fn choose_multiple<R>(&self, rng: &mut R, amount: usize) -> SliceChooseIter<Self, Self::Item>
    where
        R: Rng + ?Sized,
    {
        let amount = ::core::cmp::min(amount, self.len());
        SliceChooseIter {
            slice: self,
            _phantom: Default::default(),
            indices: index::sample(rng, self.len(), amount).into_iter(),
        }
    }
```

Reply:
Oh I see, I looked within `index::sample`. Missed the outer `min`. So I guess we can do w/o the min calls in the current fixed code, but let's keep them for being more explicit and clear